libass/ass_render.c

```diff
@@ -508,7 +508,7 @@ static bool quantize_transform(double m[3][3], ASS_Vector *pos,
     //  = m_zx * (x + sign(m_zx) * dx) + m_zy * (y + sign(m_zy) * dy) + z0.
 
     // D(f)--absolute value of error in quantity f
-    // as function of error in matrix coefficients, i. e. D(m_??).
+    // as function of error in matrix coefficients, i. e. D(m_kl).
```

Comment:
I'd prefer symbols instead of letters here (`m_**`, `m_##`, `m_@@` or `m_$$`). It's not good to introduce new identifiers without definition, especially if that definition isn't quite trivial: `k, l = {x, y, z}`.

Reply:
I find it harder to understand those two identical non-alphabetic symbols¹, instead of two different labels. Personally I'd assume any unbound label to be a placeholder for any valid value, especially for a common enumerators like`i, j, k, l`, but this might also be a non-universal convention like the `**, @@, …`.
Perhaps it's best to just add `for all k, l from {x, y, z}` to the first/all use(s)?

¹: In fact while context made the intent clear in the original `??` case, I thought the `??` would be used as shell globs, where `?` matches any single character.

Reply:
Just to add another perspective:

> Personally I'd assume any unbound label to be a placeholder for any valid value, especially for a common enumerators like`i, j, k, l`

Personally, I’m not used to seeing enumerators go as far as ℓ (and k is borderline), so I’d be confused by a new, unexplained mention of something involving it. Even the new version with the explainer comment feels kinda weird to me.

I was going to say that I find it an improvement that this now explicitly describes exactly which index combinations are intended… but is that actually right here? meaningful? After another few glances at the surrounding text, this one sentence seems to mean to say:

> Let’s write D(thing) for the absolute error in the thing’s value that we’ll get after all transformations are done and reverted, relative to the value the thing would have in an ideal rendering. We will now find reduced expressions for D(various things) that involve only various combinations of D(any element of matrix m) and no D(other things).

which

1. isn’t immediately clear from the sentence at all, whether before or after this PR;
2. doesn’t necessary involve “all” combinations of x, y, z.

Reply:
> Personally, I’m not used to seeing enumerators go as far as ℓ

In real math texts there is always shortage of letters, so I've seen _ijklmn_ and even _pqrts_ as indices.
But in this particular case `m_ij` is indeed better than `m_kl` if we go with letters.

Reply:
Now I've reworded the whole introductory paragraph, is this easier to understand and still correct?
@MrSmile: Why is `z0` a constant even though it depends on matrix coefficients? Also I don't understand what `ACCURACY` in the second touched paragraph is or why it is bit-anded with `D(y_out)`.

Reply:
I think that rewording shouldn't be done in trigraph commit.

> `i, j` are already used as variables by the function.

It's unrelated context. Also mathematicians often reuse the same letters as indices (for example, index _i_ and imaginary unit _i_ in the same expression simultaneously).

> Why is `z0` a constant even though it depends on matrix coefficients?

It depends on initial non-quantized matrix coefficients. Also it's an arbitrary scale factor that cancels out later. So there are two equivalent approaches: divide all coefficients by `z0` first and consider `z0 = 1` later, or treat `z0` as constant.

> Also I don't understand what `ACCURACY` in the second touched paragraph is or why it is bit-anded with `D(y_out)`.

`ACCURACY` is target error, and all that text is in pseudo-math notation, where `&` is simply "and" and `~` is "have the same order as".

Reply:
> I think that rewording shouldn't be done in trigraph commit.

Will split.

> > `i, j` are already used as variables by the function.
> 
> It's unrelated context. Also mathematicians often reuse the same letters as indices (for example, index _i_ and imaginary unit _i_ in the same expression simultaneously).

I've certainly seen that double use of `i` in the same expression before, but I always held the opinion, and still do, that this is poor style and should be avoided. In this case reusing `i, j` wouldn't be as bad, but we're not short on letters either, so using `k, l` is better imo and I don't see any advantage in using `i, j` instead.

> `ACCURACY` is target error, and all that text is in pseudo-math notation, where `&` is simply "and" and `~` is "have the same order as".

I'd rather use a comma instead of `&` and `≈` *(or just write out `same magnitude` if you want to stick to ASCII)* instead of `~`, which I *primarily* know as “proportional to”.

Reply:
Later on you're also using `~=`, I assume this is supposed to be an ASCII-fallback for `≈`?

Reply:
Actually, `i, j` have used in the code in exactly the same context: as vector/matrix indices. So reusing them is even better.

> which I _primarily_ know as “proportional to”.

It's not wrong too. Basically by `x ~ y` I mean `x ≈ C * y` with some constant not far from 1.

> Later on you're also using `~=`, I assume this is supposed to be an ASCII-fallback for `≈`?

Yea. Maybe we should decide that source encoding is strictly UTF-8 and permit special symbols? Math descriptions would be much better with them.

Reply:
> Actually, i, j have used in the code in exactly the same context

It's used to enumerate various things, not only the matrix.

@MrSmile , @astiob : How's the new version?